### DISCUSSION TITLE
New Toon Shader Features Discussion

### DISCUSSION THREAD

**Discussion by Geenz** [Yellow-Dog-Man/Resonite-Issues#1184] [New Features]
I'm starting this discussion thread to start collecting feedback from users for what they want from our new toon shader.

Please note that we **will not** entertain any requests to implement shaders such as poiyomi, or any other specific shader for that matter.  We may consider specific features you find useful from other shaders, but we will not be wholesale implementing these other shaders.

**Comment by TisFoolish** [comment]
This might be best as a separate post/feature request because it can apply to other shaders/materials and it's about UI and not the shader itself, but right for two different views; a Simple view that only shows the most basic features most users will commonly use, and a Full view that shows all the features and options available.

Right now the UI for the toon shader material is overwhelming, especially for new users. There is so much to it that it's quite possible for a new user to find it, see all the options, and go "I'm out" and not use it.  Having a toggle to switch from a simple view to a full view would help a lot

**Reply by Frooxius** [reply]
I'd definitely split this out. This isn't even specifically for just shaders/materials, but components in general, having an "abridged" view showing the most important parts of component and then detailed "full" view. 

I wouldn't want us to implement this specifically for just one shader, but rather have a generic approach for this.

**Reply by Geenz** [reply]
Yeah - it's something we want to do at some point for other things.

**Reply by TisFoolish** [reply]
I could make a formal feature request if you'd like, Froox. Or leave it to you to do so if you want.

**Reply by LexiBasilisk** [reply]
Yes, as Frooxius had mentioned above- split this off into its own feature request issue for later consideration, @TisFoolish.

**Comment by lxw404** [comment]
This may be a wider request as well that applies to all shaders in general:

I would certainly like to see every single texture input have its own individual `TextureOffset` and `TextureScale` parameters instead of randomly tying these to multiple maps at once since many effects depend on these being separated, especially for normal map modulation effects.

**Comment by Epimonster** [comment]
### Packing Standards
Currently the xiexie toon shader supports a PBR setup via the MetallicGlossMap. However, the MetallicGlossMap uses the channel packing standard of
R - Metallic, Ambient Occlusion
G - N/A
B - N/A
A - Smoothness

I would like to request synchronizing the new toon shader with PBS_Metallic's packing standard which is:
R - Metallic
G - Ambient Occlusion or Height
B - N/A
A - Smoothness

### Matcaps
As far as direct features the first thing that comes to mind is better matcap support in the toon shader. The ability to control the way a matcap blends with a shader would be nice, choosing whether its additive, multiplicative, etc.

Most useful though would be the ability to mask matcaps with a texture. This is a request I've heard from a ton of users, new and old alike so I'd like to suggest it. Optimally it would be a setup where white is just the matcap and with black the matcap has no influence over the texture.

In a perfect world I would love to have support for up to 4 matcaps and a colored matcap mask where each channel of the mask texture would be used to control a matcap's influence over a given area. I would advocate for supporting multiple matcaps on a material as it allows for quite a wide variety of materials to be represented in a toon styled way on one material.

**Reply by akiRAM2** [reply]
I second this.
Especially, the ability to select the matcap blend mode is essential for toon shaders.
Without it, it's very hard to achieve the proper specular effect.
![image](https://github.com/user-attachments/assets/ce908c37-f8d3-465a-86d8-a3d614ee6ecf)
![image](https://github.com/user-attachments/assets/8449a5db-ceb6-4ab6-9457-c2564c451e28)
![image](https://github.com/user-attachments/assets/b8262725-097f-4a56-87a7-fdf0f2ecc3ae)

**Comment by PointerOffset** [comment]
I'd be interested to see some type of screen-space pattern or mask that can be applied to shading to further stylize shadows/highlights. In particular I'm thinking of cross-hatching or screen print dots and similar stylization effects.

I'm not really sure of the best way to implement this. I'll have to look up some shaders with examples but I can imagine:
 - An enum with a selection of shading styles paired with some controls for the size/shape of the shading pattern. Things like thickness of individual dots/hatch marks in the pattern or pattern density along the shadow ramp.
 
 - A texture from the user that lets them define the pattern style along the Shadow Ramp. This would mean people could create all kinds of interesting shading patterns but I don't know if we'd have a harder time controlling the appearance compared to a few options with some sliders.
 
 Edit: I came across a handful of shader examples showing the types of styles I have in mind: https://spite.github.io/sketch/

**Comment by JackTheFoxOtter** [comment]
There's definitely a couple of important parts from Xixie that should be supported too. Mainly custom shadow ramps and outlines / outline masks. One addition I'd like to have for that is a way to specify variying thickness of the outline on different parts of a model, using an additional texture input.

I would also definitely prefer the shader to not break when there's more than like, 4 lights affecting it.

**Reply by lxw404** [reply]
Additional modes for outlines would also be useful, specifically `Directional` and `Offset with Camera Z` which have various stylistic approaches.

**Comment by lxw404** [comment]
Not to say that I'm biased, but stencil properties w/ ZTest would be a welcome addition.  One major thing that this would enable would be features of drawn characters such as eyebrows which should draw above other parts of the character, along with other advanced stencil operations.

Another often heard complaint is that there's no easy way to do glitter, though I'm not sure I'd consider it a staple of toon-specific shading rather than its own thing.

**Comment by KisaragiEffective** [comment]
I stuck in following situation during importing assets or avatars that built for VRChat (or whatever using lilToon):

1. "2nd texture" toggle -> fixed by having two texture where only difference is the texture is drawn
2. jem material -> ignored, as it does not seem that Resonite support it
3. Specifying ShadowLamp on XiexeToon -> specified new texture that has gradation from white to beige.

and why those are problem (for me):
1. is not good for VRAM, especially original "2nd texture" has small resolution
2. would be better to supported anyway, wouldn't it?
3. compound reason:
	1. lilToon does not have it
	2. default ShadowLamp do *not* yield similar result compared from lilToon.

in addition to that, lacking support for masking matcap on XiexeToon sometimes become a issue for porting.

A Japanese community member has written [article](https://note.com/akiram_vr/n/n5e55290e2cee) about "Frequently Asked Question" for XiexeToon which mentions some of mentioned above (and part of lilToon features).

**Reply by KisaragiEffective** [reply]
related: #3258 .

**Comment by Redd56** [comment]


**Reply by lxw404** [reply]
Most "toon" shaders begin with physically based properties, and then modulate/stylize those for artistic purposes.  Subsurface scattering allows for backlighting, which is a very important property for stylized foliage.  Additionally, a matcap is one of the most important stylization options for getting cheap and consistent fake lighting.  It's probably the most widely used option on "toon" shaders, so neither of these are out of place.

**Reply by Redd56** [reply]
Huh didnt know that about matcap. Tho the scattering does make sence now that you mention it. I more think of toon shaders in terms of drawn flat art toon and dont really think about stylized styles like say rares sea of thieves much

**Reply by Redd56** [reply]
im killing younger me for this opinion. matcaps are goated. and subsurface scattering is like goated as well though not as used

**Comment by EnergyBun** [comment]
From what I have seen LilToon is the most popular Unity shader.  I don't know anything else about it, however, based on this simple fact, the new shader for Resonite needs to do all the things LilToon does in the exact same way in order to bring in users.

**Reply by Frooxius** [reply]
It would help us to understand what makes the shader so popular - that way we can focus on giving people the features they need the most.

I don't believe things are necessarily that black and white and that it needs to be exact copy of this shader. Shaders, like any thing often have features that are very commonly used and ones that aren't used as commonly - ones that most people might not care if they were missing or handled differently.

Replicating shader exactly is a lot more difficult than understanding what makes it good and offering the most popular features.

**Reply by Redd56** [reply]
most common is def the outline. most avatars that use toon that i see in vrc (tho not resonite cause not everyone knows about the outline) use the toon outline. even some without toon shading use it.

**Reply by ko-tengu** [reply]
Color masks, Alpha masks, matcap masks, multiple matcaps, decal layers, and the ability to set minimum and maximum light levels, would probably be the main picks of functions liltoon/poiyomi do that Resonite can't with it's current shader.

**Reply by Frooxius** [reply]
Thank you for more context!
